File: webapp/Locust/Scripts/app/services/TestScenarioService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('TestScenarioService', ['$resource', function ($resource) {

        var noop = angular.noop;
        var resource = $resource('/api/TestScenario/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                Update: {
                    method: 'PUT'
                },
                Restore: {
                    method: 'POST'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Enable: {
                    method: 'GET',
                    isArray: false
                },
                GetLastTestScenario: {
                    method: 'GET',
                    isArray: false
                },
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetLastOne: {
                    method: 'GET',
                    isArray: false
                },

                GetAll: {
                    method: 'GET',
                    isArray: true
                },

                UpdateNumber: {
                    method: 'PUT'
                },

                GetProject: {
                    method: 'GET',
                    isArray: false
                },
                GetRequirement: {
                    method: 'GET',
                    isArray: false
                },
                GetPtojectRequirement: {
                    method: 'GET',
                    isArray: false
                },
                AddChangeLog: {
                    method: 'GET',
                    isArray: false
                },
                TestScenarioChangeLogs: {
                    method: 'GET',
                    isArray: true
                },
                GetRequirementForTs: {
                    method: 'GET',
                    isArray: false
                }
            }),
            self = this;
        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };
        self.Get = function (id, success, error) {
            return resource.Get({ action: 'Get', id: id }, success || noop, error || noop);
        };
        self.GetLastOne = function (idReq, creator, date, success, error) {
            return resource.GetLastOne({ action: 'GetLastOne', idReq: idReq, creator: creator, date: date }, success || noop, error || noop);
        };
        self.GetRequirement = function (reqId, success, error) {
            return resource.GetRequirement({ action: 'GetRequirement', reqId: reqId }, success || noop, error || noop);
        };
        self.Delete = function (id, success, error) {
            return resource.Delete({ action: 'Delete', id: id }, success || noop, error || noop);
        };
        self.Enable = function (id, success, error) {
            return resource.Enable({ action: 'Enable', id: id }, success || noop, error || noop);
        };
        self.GetLastTestScenario = function (creator, date, success, error) {
            return resource.GetLastTestCase({ action: 'GetLastTestScenario', creator: creator, date: date }, success || noop, error || noop)
        };
        self.GetProject = function (idts, success, error) {
            return resource.GetProject({ action: 'GetProject', idts: idts }, success || noop, error || noop)
        };
        self.AddChangeLog = function (id, success, error) {
            return resource.AddChangeLog({ action: 'AddChangeLog', id: id }, success || noop, error || noop);
        };
        self.Restore = function (entity, success, error) {
            return resource.Restore({ action: 'Restore' }, entity, success || noop, error || noop);
        };
        self.GetProjectRequirement = function (idreq, success, error) {
            return resource.GetProject({ action: 'GetProjectRequirement', idreq: idreq }, success || noop, error || noop)
        };
        self.TestScenarioChangeLogs = function (id, success, error) {
            return resource.TestScenarioChangeLogs({ action: 'TestScenarioChangeLogs', id: id }, success || noop, error || noop);
        };
        self.Save = function (entity, success, error) {
            return resource.Save({ action: 'Save' }, entity, success || noop, error || noop);
        };

        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };
        self.UpdateNumber = function (entity, idReq, success, error) {
            return resource.UpdateNumber({ action: 'UpdateNumber', idReq: idReq }, entity, success || noop, error || noop);
        };
        self.GetRequirementForTs = function (tsId, success, error) {
            return resource.GetRequirementForTs({ action: 'GetRequirementForTs', tsId: tsId }, success || noop, error || noop);
        };
    }]);
})(angular);

File: webapp/Locus.Core/Services/TestScenarioService.cs
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface ITestScenarioService
    {
        TestScenario Save(TestScenario testScenario, string user);

        TestScenario Update(TestScenario testScenario, string user);

        TestScenario Delete(int idTestScenario);

        TestScenario Enable(int idTestScenario);

        List<TestScenario> GetAll();

        TestScenario Get(int idTestScenario);

        TestScenario GetLastOne(int idReq, string creator, string date);

        TestScenario GetLastTestScenario(string creator, string date);

        TestScenario UpdateNumber(TestScenario testScenario, int idReq);

        Project GetProject(int idts);

        Requirement GetRequirement(int idts);

        Project GetProjectRequirement(int idreq);

        ChangeLog AddChangeLog(int id, string user);

        List<ChangeLog> TestScenarioChangeLogs(int id);

        ChangeLog Restore(ChangeLog change_log, string user);

        Requirement GetRequirementForTs(int tsId);
    }

    public class TestScenarioService : ITestScenarioService
    {
        private readonly ITestScenarioRepository _testScenarioRepository;
        private readonly ITestResultRepository _testResultRepository;

        //Constructor of TestCaseService and initialize testCaseRepository
        public TestScenarioService(ITestScenarioRepository testScenarioRepository)
        {
            _testScenarioRepository = testScenarioRepository;
        }

        public TestScenarioService(ITestScenarioRepository testScenarioRepository, ITestResultRepository testResultRepository)
        {
            _testScenarioRepository = testScenarioRepository;
            _testResultRepository= testResultRepository;
        }

        public ChangeLog AddChangeLog(int id, string user)
        {
            return _testScenarioRepository.AddChangeLog(id, user);
        }

        public TestScenario Delete(int idTestScenario)
        {
            return _testScenarioRepository.Delete(idTestScenario);
        }

        public TestScenario Enable(int idTestScenario)
        {
            return _testScenarioRepository.Enable(idTestScenario);
        }

        public TestScenario Get(int idTestScenario)
        {
            return _testScenarioRepository.Get(idTestScenario);
        }

        public List<TestScenario> GetAll()
        {
            return _testScenarioRepository.GetAll();
        }

        public TestScenario GetLastOne(int idReq, string creator, string date)
        {
            return _testScenarioRepository.GetLastOne(idReq, creator, date);
        }

        public TestScenario GetLastTestScenario(string creator, string date)
        {
            return _testScenarioRepository.GetLastTestScenario(creator, date);
        }

        public Project GetProject(int idts)
        {
            return _testScenarioRepository.GetProject(idts);
        }

        public Project GetProjectRequirement(int idreq)
        {
            return _testScenarioRepository.GetProjectRequirement(idreq);
        }

        public Requirement GetRequirement(int idts)
        {
            return _testScenarioRepository.GetRequirement(idts);
        }

        public Requirement GetRequirementForTs(int tsId)
        {
            return _testScenarioRepository.GetRequirementForTs(tsId);
        }

        public ChangeLog Restore(ChangeLog change_log, string user)
        {
            return _testScenarioRepository.Restore(change_log, user);
        }

        public TestScenario Save(TestScenario testScenario, string user)
        {
            return _testScenarioRepository.Save(testScenario, user);
        }

        public List<ChangeLog> TestScenarioChangeLogs(int id)
        {
            return _testScenarioRepository.TestScenarioChangeLogs(id);
        }

        public TestScenario Update(TestScenario testScenario, string user)
        {
            _testResultRepository.UpdateTitles(testScenario.Test_Scenario_Id, testScenario.Title, "TS");
            return _testScenarioRepository.Update(testScenario, user);
        }

        public TestScenario UpdateNumber(TestScenario testScenario, int idReq)
        {
            return _testScenarioRepository.UpdateNumber(testScenario, idReq);
        }
    }
}
